5.5 ps bug1368530

ihanick opened the pull request.

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,16 @@
+# Test that there is no conflict between the I_S plugin and the changed page bitmap option names (LP:1105726)
```
Leftover from another testcase?

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,16 @@
+--source include/have_innodb.inc
+--source include/have_debug.inc
+
+# If the bug is not present, server starts
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,16 @@
+# If the bug is not present, server starts
+DROP DATABASE test;CREATE DATABASE test;
+USE test;
+SET @@GLOBAL.innodb_track_changed_pages=FALSE;
```
This appears to test something else than what needed fixing. Here we start if var=FALSE, set it to FALSE again (succeed), then TRUE (succeed but should have failed instead), then FALSE again (fail).

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,16 @@
+--Error ER_WRONG_VALUE_FOR_VAR
+SET @@GLOBAL.innodb_track_changed_pages=FALSE;
+SELECT @@GLOBAL.innodb_track_changed_pages;
+PURGE CHANGED_PAGE_BITMAPS BEFORE 0;
```
Why?

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,16 @@
+SELECT @@GLOBAL.innodb_track_changed_pages;
+PURGE CHANGED_PAGE_BITMAPS BEFORE 0;
+
+--source include/restart_mysqld.inc
```
Why?

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -12476,6 +12476,69 @@ innodb_change_buffering_update(
+#ifdef UNIV_DEBUG
+static
+int
+innodb_track_changed_pages_validate(
```
This and _update implement incorrect logic. The tracking subsys should never be initialised at runtime, the only allowed thing is to temporarily stop and restart it.

laurynas-biveinis commented:
```
- The 5.5 version of the fix cannot happen on the 5.5 trunk last
  revision, because it will be unmergeable to 5.6 then (it will
  try to merge all the unmerged 5.5 revisions, twenty or so of
  them). So the fix must be based on an earlier commit of 5.5,
  which is fully merged to 5.6. One way to get such base revision
  is  `git rev-list percona-5.5 ^percona-5.6 --first-parent --topo-order | tail -1`^

- The fix commit and the testcase commits should be merged.

- The commit message (a single one after merging the commits)
  should be descriptive: a "title" as the first line (bug id and
  shortly what is it about) for git log --oneline, then analysis
  and fix description.

- A link to a Jenkins run should be provide with a PR (ping PeterS
  if you don't have permissions there)
```

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1155/

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,50 @@
+--Error ER_WRONG_VALUE_FOR_VAR
+SET @@GLOBAL.innodb_track_changed_pages=TRUE;
+SELECT @@GLOBAL.innodb_track_changed_pages;
+PURGE CHANGED_PAGE_BITMAPS BEFORE 0;
```
Redundant?

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,50 @@
+
+--enable_reconnect
+--echo 1st restart
+--exec echo "restart:--innodb-data-home-dir=$TMPDATADIR --innodb_log_group_home_dir=$TMPDATADIR --innodb_track_changed_pages=1 --innodb_changed_pages=FORCE" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
```
"--innodb_changed_pages=FORCE" bit redundant?

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,50 @@
+SET @@GLOBAL.innodb_track_changed_pages=TRUE;
+SELECT @@GLOBAL.innodb_track_changed_pages;
+
+--remove_files_wildcard $TMPDATADIR ib_modified_log_*.xdb
```
Redundant?

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,50 @@
+
+--remove_files_wildcard $TMPDATADIR ib_modified_log_*.xdb
+
+PURGE CHANGED_PAGE_BITMAPS BEFORE 0;
```
Redundant?

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -215,6 +215,11 @@ static char*	innodb_version_str = (char*) INNODB_VERSION_STR;
 
+
+#ifdef UNIV_DEBUG
+static bool     srv_track_changed_pages_was_active = FALSE;
```
Needs a comment. If InnoDB coding style allows, can be moved as static var into innodb_track_changed_pages_validate.

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -12476,6 +12481,43 @@ innodb_change_buffering_update(
+		return 1;
+	}
+
+	if (srv_track_changed_pages || srv_track_changed_pages_was_active) {
```
Here or in the function header (or maybe adding a comment to srv_track_changed_pages_was_active will resolve) needed a short allowed option manipulation description comment

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1165/
- "PURGE CHANGED_PAGE_BITMAPS BEFORE 0;" is used to crash mysqld with the same stack trace as described in bug (unmodified version)
- clean up for test case, only 1(default)=>0=>1 transition is tested.
- updated validation function comment
- removed global changed pages init flag => static variable inside function
  Static variables inside functions are not frequently used inside innodb, but there are several places in vanilla mysql and percona server 5.5
  5.7.8: buf/buf0buf.cc:507 buf/buf0flu.cc:2428 fil/fil0fil.cc:240
  in PS 5.5 innodb buf/buf0buf.c:485 trx/trx0purge.c:670 page/page0cur.c:63

laurynas-biveinis commented on mysql-test/suite/innodb/t/percona_changed_page_bmp_1368530.test:
```diff
@@ -0,0 +1,34 @@
+--enable_reconnect
+--echo 2nd restart
+--exec echo "restart" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
+--source include/wait_until_connected_again.inc
```
If you removed tmpdir / rmdir, then we need to do RESET BITMAPS in order not to leave stray xdb around

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -12476,6 +12476,46 @@ innodb_change_buffering_update(
+						for update function */
+	struct st_mysql_value*		value)	/*!< in: incoming bool */
+{
+	static bool     enabled_on_startup = FALSE;
```
s/FALSE/false and below s/TRUE/true (it's bool not ibool or my_bool)

ihanick commented:
http://jenkins.percona.com/job/percona-server-5.5-param/1166/
- cleanup for ib_modified_log_1_0.xdb
- using c++ true/false instead TRUE/FALSE

laurynas-biveinis commented:
This is OK, please merge to 5.6.
The 2nd server restart in the testcase can use --source include/restart_mysqld.inc in 5.6

ihanick commented:
5.6 PR: https://github.com/percona/percona-server/pull/206
